cmd/soroban-rpc/internal/feewindow/feewindow.go

```diff
@@ -0,0 +1,178 @@
+package feewindow
+
+import (
+	"io"
+	"slices"
+	"sync"
+
+	"github.com/stellar/go/ingest"
+	"github.com/stellar/go/xdr"
+
+	"github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/ledgerbucketwindow"
+)
+
+type FeeDistribution struct {
+	Max   uint64
+	Min   uint64
+	Mode  uint64
+	P10   uint64
+	P20   uint64
+	P30   uint64
+	P40   uint64
+	P50   uint64
+	P60   uint64
+	P70   uint64
+	P80   uint64
+	P90   uint64
+	P95   uint64
+	P99   uint64
+	Count uint64
+}
+
+type FeeWindow struct {
+	lock          sync.RWMutex
+	totalFeeCount uint64
+	feesPerLedger *ledgerbucketwindow.LedgerBucketWindow[[]uint64]
+	distribution  FeeDistribution
+}
+
+func NewFeeWindow(retentionWindow uint32) *FeeWindow {
+	window := ledgerbucketwindow.NewLedgerBucketWindow[[]uint64](retentionWindow)
+	return &FeeWindow{
+		totalFeeCount: 0,
+		feesPerLedger: window,
+	}
+}
+
+func (fw *FeeWindow) AppendLedgerFees(fees ledgerbucketwindow.LedgerBucket[[]uint64]) error {
+	fw.lock.Lock()
+	defer fw.lock.Unlock()
+	_, err := fw.feesPerLedger.Append(fees)
+	if err != nil {
+		return err
+	}
+
+	var allFees []uint64
+	for i := uint32(0); i < fw.feesPerLedger.Len(); i++ {
+		allFees = append(allFees, fw.feesPerLedger.Get(i).BucketContent...)
+	}
+	fw.distribution = computeFeeDistribution(allFees)
+
+	return nil
+}
+
+func computeFeeDistribution(fees []uint64) FeeDistribution {
+	if len(fees) == 0 {
+		return FeeDistribution{}
+	}
+	slices.Sort(fees)
+	mode := fees[0]
+	lastVal := fees[0]
+	maxRepetitions := 0
+	localRepetitions := 0
+	for i := 1; i < len(fees); i++ {
+		if fees[i] == lastVal {
+			localRepetitions += 1
+			continue
+		}
+
+		// new cluster of values
+
+		if localRepetitions > maxRepetitions {
+			maxRepetitions = localRepetitions
+			mode = fees[i]
+		}
+		lastVal = fees[i]
+		localRepetitions = 0
+	}
+	count := uint64(len(fees))
+	// nearest-rank percentile
+	percentile := func(p uint64) uint64 {
+		// ceiling(p*count/100)
+		kth := ((p * count) + 100 - 1) / 100
+		return fees[kth-1]
+	}
```

Comment:
I find [this implementation](https://github.com/montanaflynn/stats/blob/v0.7.1/percentile.go#L53-L86) easier to understand

Reply:
I considered using that but it's less efficient (it uses floats and we would need to sort the list multiple times for the other computations).

Maybe it's not worth optimizing it though ...

Reply:
We can use that implementation for testing though

Reply:
@tamirms that's a bug, it should be:

```suggestion
				sorobanInclusionFees = append(sorobanInclusionFees, inclusionFee)
```

Fixed

Reply:
I used github.com/montanaflynn/stats for testing. So, while I was at it, I created a benchmark:

```
goos: darwin
goarch: arm64
pkg: github.com/stellar/soroban-rpc/cmd/soroban-rpc/internal/feewindow
BenchmarkComputeFeeDistribution
BenchmarkComputeFeeDistribution/computeFeeDistribution
BenchmarkComputeFeeDistribution/computeFeeDistribution-20         	 8273469	       137.5 ns/op
BenchmarkComputeFeeDistribution/alternativeComputeFeeDistribution
BenchmarkComputeFeeDistribution/alternativeComputeFeeDistribution-20         	  362680	      3122 ns/op
PASS
```

The implementation is 20x faster. Not that, spending 6 microseconds (two computations) every ledger was going to cause any havoc ...